mrp_unbuild_subcontracting/models/stock_picking.py

```diff
@@ -0,0 +1,110 @@
+from datetime import timedelta
+
+from dateutil.relativedelta import relativedelta
+
+from odoo import fields, models
+from odoo.osv.expression import OR
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    subcontracted_unbuild_ids = fields.One2many(
+        "mrp.unbuild", "picking_id", readonly=True, string="Suncontracted unbuilds"
+    )
+
+    def _prepare_subcontract_unbuild_vals(self, subcontract_move, bom):
+        subcontract_move.ensure_one()
+        product = subcontract_move.product_id
+        group = self.env["procurement.group"].create(
+            {
+                "name": self.name,
+                "partner_id": self.partner_id.id,
+            }
+        )
+        warehouse = self._get_warehouse(subcontract_move)
+        vals = {
+            "company_id": subcontract_move.company_id.id,
+            "product_id": product.id,
+            "product_uom_id": subcontract_move.product_uom.id,
+            "procurement_group_id": group.id,
+            "bom_id": bom.id,
+            "location_id": subcontract_move.picking_id.partner_id.with_company(
+                subcontract_move.company_id
+            ).property_stock_subcontractor.id,
+            "location_dest_id": subcontract_move.picking_id.partner_id.with_company(
+                subcontract_move.company_id
+            ).property_stock_subcontractor.id,
+            "product_qty": subcontract_move.product_uom_qty,
+            "picking_type_id": warehouse.subcontracting_type_id.id,
+            "date_planned_start": subcontract_move.date
+            - relativedelta(days=product.produce_delay),
+            "picking_id": self.id,
+            "is_subcontracted": True,
+        }
+        return vals
+
+    def _subcontracted_produce_unbuild(self, subcontract_details):
+        self.ensure_one()
+        for move, bom in subcontract_details:
+            unbuild = (
+                self.env["mrp.unbuild"]
+                .with_company(move.company_id)
+                .create(self._prepare_subcontract_unbuild_vals(move, bom))
+            )
+            unbuild.date_planned_finished = (
+                move.date
+            )  # Avoid to have the picking late depending on the unbuild order
+            self.subcontracted_unbuild_ids |= unbuild
+
+    def _action_done(self):
+        res = super(StockPicking, self)._action_done()
+        for picking in self:
+            unbuilds_to_done = (
+                picking.subcontracted_unbuild_ids._subcontracting_filter_to_done()
```

Comment:
Why don't just filter on the go with a lambda function? I only see this function used here.